### ISSUE TITLE
Improve configuration

### ISSUE THREAD

**Issue by ryuukk** [ufbx/ufbx#153]
Hello, i'm trying to use this library in my project, however, i encounter few issues

- defining: `UFBX_USE_SSE 0` or `UFBX_WASM_32BIT` should not include `xmmintrin.h` and `emmintrin.h`

- consider: add define to disable stdio (file loading)

- consider: document or add define to disable/override usage from libc (see stdb_image.h for inspiration)

As it is, can't be used for WASM without emscripten

I'll try to provide a PR if i have time

**Comment by bqqbarbhg** [comment]
Heya!

The configuration defines are existence-based as in some systems it's more cumbersome to define values. Thus you should be fine as long as you don't define `UFBX_USE_SSE`, as it's only enabled when the architecture is detected to be x86-64:

```c
#if !defined(UFBX_STANDARD_C) && (defined(_MSC_VER) && defined(_M_X64)) || ((defined(__GNUC__) || defined(__clang__)) && defined(__x86_64__)) || defined(UFBX_USE_SSE)
```

If you are hitting this issue you may have some cross-compilation issue with WASM where the `__x86_64__` define spills from the host OS.

The `UFBX_USE_SSE` define is meant specifically only for 32-bit x86 architecture, since it does support SSE, but only in specific processors. The define lets users opt into the requirement. I guess there could be a `UFBX_NO_SSE` define (or something adjacent) that disables it from x86-64, but it might be simpler to just have something like `UFBX_NO_INTRINSICS` which would disable SSE and other CPU-specific things like `ufbxi_div128()` or future NEON support.

For now, just not defining `UFBX_USE_SSE` should work, unless you specifically want to not have SSE on x86-64. You can do that currently only by defining `UFBX_STANDARD_C` which disables all non-standard features (but also makes the library less optimal as it messes with inline/noinline directives).

---

For the libc dependency: ufbx uses libc a fair bit more than for example `stb_image.h`. Currently the full libc is expected as a hard dependency. Disabling the `FILE` API of `stdio.h` would be reasonable to do, but ufbx needs `vsnprintf()` etc. from the header, I'm not sure if this is an issue.

I guess the cleanest way to approach this would be to allow users to specify `UFBX_NO_LIBC`. Defining this would disable _all_ `#include` statements from the source and the header. The configuration would look something like this on the users end for example:

```c
#define UFBX_NO_LIBC
#define UFBX_NO_STDIO

// Type defines
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>

// Compiled in, no libc dependency
#include <stdarg.h>

// Custom libc implementation for subset that is needed
#include "my_libc.h"

#define memcpy my_memcpy
#define strlen my_strlen
#define stdtof my_strtof
// ... etc
```

---

I'll see if I can get this working, no need to do a PR since I have a solid idea for how I want this to be done. Thanks for the issue!

**Comment by bqqbarbhg** [comment]
It still requires a bit of polish but the branch `no-libc` from the PR #154 has proper support for building without libc now. The documentation is still a bit lacking and it needs a bit more testing, but it seems quite stable now.

To compile without libc, all you need to do is have `UFBX_NO_LIBC` defined when building `ufbx.c`. This will prevent ufbx from importing any standard headers, except `stdint.h`, `stdbool.h`, `stddef.h`, `stdarg.h`. These are generally available, even when not using libc, for example when building with `--target=wasm32`. Note that you might need to build with `-mbulk-memory` as otherwise Clang will insert `memcpy()` and `memset()` calls.

Even though the PR moved a lot of libc functionality into ufbx (which is something I have wanted to do for a while, due to eg. `strtod()` precision issues on some platforms and locale issues), there is still a whole lot of functionality that ufbx needs.

Here is a list of all the functions that ufbx needs, and ways to avoid having to define them:

Either add [extra/ufbx_math.c](https://github.com/ufbx/ufbx/blob/no-libc/extra/ufbx_math.c) to the build or define the following functions:
```c
double ufbx_sqrt(double x);
double ufbx_sin(double x);
double ufbx_cos(double x);
double ufbx_tan(double x);
double ufbx_asin(double x);
double ufbx_acos(double x);
double ufbx_atan(double x);
double ufbx_atan2(double y, double x);
double ufbx_pow(double x, double y);
double ufbx_fmin(double a, double b);
double ufbx_fmax(double a, double b);
double ufbx_fabs(double x);
double ufbx_copysign(double x, double y);
double ufbx_nextafter(double x, double y);
double ufbx_rint(double x);
double ufbx_ceil(double x);
int ufbx_isnan(double x);
```

Either add [extra/ufbx_libc.c](https://github.com/ufbx/ufbx/blob/no-libc/extra/ufbx_libc.c) to the build or define the following functions:
```c
size_t ufbx_strlen(const char *str);
void *ufbx_memcpy(void *dst, const void *src, size_t count);
void *ufbx_memmove(void *dst, const void *src, size_t count);
void *ufbx_memset(void *dst, int ch, size_t count);
const void *ufbx_memchr(const void *ptr, int value, size_t count);
int ufbx_memcmp(const void *a, const void *b, size_t count);
int ufbx_strcmp(const char *a, const char *b);
int ufbx_strncmp(const char *a, const char *b, size_t count);
```

Either define `UFBX_NO_MALLOC` to disable the default memory allocator or define the following functions:
```c
void *ufbx_malloc(size_t size);
void *ufbx_realloc(void *ptr, size_t old_size, size_t new_size);
void ufbx_free(void *ptr, size_t old_size);
```

Either define `UFBX_NO_STDIO` to disable the default file I/O or define the following functions:
```c
void *ufbx_stdio_open(const char *path, size_t path_len);
size_t ufbx_stdio_read(void *file, void *data, size_t size);
bool ufbx_stdio_skip(void *file, size_t size);
uint64_t ufbx_stdio_size(void *file);
void ufbx_stdio_close(void *file);
```

**Comment by bqqbarbhg** [comment]
Landed in `v0.15.0`
